<a href="https://colab.research.google.com/github/Apur52027/deep_learning/blob/main/Detection_of_autism_spectrum__DL_Assignment_03_Question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# DL Assignment 03

**Name:Vupal Chandra Roy**

**Course Email:apur52027@gmail.com**  


## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।

# General Instruction

You must choose your own dataset.

The dataset must:

Be a supervised learning dataset (Regression or Binary Classification)

Contain at least 300 samples

Have at least 2 input features

Be in CSV format

You are NOT allowed to use Dataset or DataLoader.

You must implement everything manually.

# Question 01: [ Marks 05 ]

## Dataset Preparation

## Using your chosen dataset:

Load the dataset.

Perform necessary preprocessing:

Handle missing values (if any)

Encode categorical variables (if necessary)

Feature scaling (if needed)

Separate features (X) and target (y).

Convert them into NumPy arrays.

Convert them into PyTorch tensors.

Split into training and testing sets.

Clearly explain each preprocessing decision.

In [ ]:
import numpy as np
import pandas as pd
import torch
data = pd.read_csv(' https://raw.github.com/Apur52027/Dataset_ML/main/autism_screening.xls ')
data.head()

,A1_Score,A2_Score,A3_Score,A4_Score,A5_Score,A6_Score,A7_Score,A8_Score,A9_Score,A10_Score,...,gender,ethnicity,jundice,austim,contry_of_res,used_app_before,result,age_desc,relation,Class/ASD
0,1,1,1,1,0,0,1,1,0,0,...,f,White-European,no,no,United States,no,6.0,18 and more,Self,NO
1,1,1,0,1,0,0,0,1,0,1,...,m,Latino,no,yes,Brazil,no,5.0,18 and more,Self,NO
2,1,1,0,1,1,0,1,1,1,1,...,m,Latino,yes,yes,Spain,no,8.0,18 and more,Parent,YES
3,1,1,0,1,0,0,1,1,0,1,...,f,White-European,no,yes,United States,no,6.0,18 and more,Self,NO
4,1,0,0,0,0,0,0,1,0,0,...,f,?,no,no,Egypt,no,2.0,18 and more,?,NO


In [ ]:
for col in data.columns:
    data[col] = data[col].replace('?',np.nan)

In [ ]:
data.shape

(704, 21)

In [ ]:
data.drop(columns=["age_desc"], inplace=True)

In [ ]:
data.describe().T

,count,mean,std,min,25%,50%,75%,max
A1_Score,704.0,0.721591,0.448535,0.0,0.0,1.0,1.0,1.0
A2_Score,704.0,0.453125,0.498152,0.0,0.0,0.0,1.0,1.0
A3_Score,704.0,0.457386,0.498535,0.0,0.0,0.0,1.0,1.0
A4_Score,704.0,0.495739,0.500337,0.0,0.0,0.0,1.0,1.0
A5_Score,704.0,0.498580,0.500353,0.0,0.0,0.0,1.0,1.0
A6_Score,704.0,0.284091,0.451301,0.0,0.0,0.0,1.0,1.0
A7_Score,704.0,0.417614,0.493516,0.0,0.0,0.0,1.0,1.0
A8_Score,704.0,0.649148,0.477576,0.0,0.0,1.0,1.0,1.0
A9_Score,704.0,0.323864,0.468281,0.0,0.0,0.0,1.0,1.0
A10_Score,704.0,0.573864,0.494866,0.0,0.0,1.0,1.0,1.0


In [ ]:
print("Total no. of missing values in Adults's dataset     : ",data.isnull().sum().sum())

Total no. of missing values in Adults's dataset     :  192


In [ ]:
data.isnull().sum()

,0
A1_Score,0
A2_Score,0
A3_Score,0
A4_Score,0
A5_Score,0
A6_Score,0
A7_Score,0
A8_Score,0
A9_Score,0
A10_Score,0


In [ ]:
from sklearn.impute import SimpleImputer
imputer_mode = SimpleImputer(missing_values=np.nan, strategy='most_frequent')

data.ethnicity = imputer_mode.fit_transform(data.ethnicity.values.reshape(-1,1))[:,0]
data.relation = imputer_mode.fit_transform(data.relation.values.reshape(-1,1))[:,0]


imputer_mode = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
data.age = imputer_mode.fit_transform(data.age.values.reshape(-1,1))[:,0]

In [ ]:
print("No of missing values in Adults's dataset      = ",data.isnull().sum().sum())

No of missing values in Adults's dataset      =  0


In [ ]:

data.rename(columns = {'age':'Age_Mons'}, inplace = True)

data['Age_Mons'] = data['Age_Mons']*12

In [ ]:
data['ethnicity']

,ethnicity
0,White-European
1,Latino
2,Latino
3,White-European
4,White-European
...,...
699,White-European
700,Hispanic
701,White-European
702,South Asian


In [ ]:
data['ethnicity'].value_counts()

,count
ethnicity,
White-European,328
Asian,123
Middle Eastern,92
Black,43
South Asian,36
Others,30
Latino,20
Hispanic,13
Pasifika,12


In [ ]:

data['ethnicity'] = data['ethnicity'].replace('Others','others')

In [ ]:
data['ethnicity'].value_counts()

,count
ethnicity,
White-European,328
Asian,123
Middle Eastern,92
Black,43
South Asian,36
others,31
Latino,20
Hispanic,13
Pasifika,12


In [ ]:
target = 'Class/ASD'
data[target].value_counts()

,count
Class/ASD,
NO,515
YES,189


In [ ]:
#Encode categorical variables
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
data['Class/ASD'] = le.fit_transform(data['Class/ASD'])
data[target].value_counts()

,count
Class/ASD,
0,515
1,189


In [ ]:
catagori_column = [col for col in data.columns if data[col].dtype == 'object']
num_column = [col for col in data.columns if data[col].dtype != 'object']
catagori_column

['gender',
 'ethnicity',
 'jundice',
 'austim',
 'contry_of_res',
 'used_app_before',
 'relation']

In [ ]:
data = pd.get_dummies(data, columns=catagori_column)

In [ ]:
#Separate features (X) and target (y).
X = data.drop(columns=[target])
y = data[target]


In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler =  StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [ ]:
#Convert them into NumPy arrays.
X_train_np = np.array(X_train)
X_test_np= np.array(X_test)
y_train_np= np.array(y_train)
y_test_np= np.array(y_test)




In [ ]:
#Convert them into PyTorch tensors.

X_train_tensor = torch.tensor(X_train_np, dtype=torch.float32)
X_test_tensor = torch.tensor(X_test_np, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train_np, dtype=torch.float32)
y_test  = torch.tensor(y_test_np, dtype=torch.float32)




# **Write** Answer 01:


# Question 02: [ Marks 20 ]

## Design a neural network using nn.Module.

### The model must contain:

Input layer

At least one hidden layer

Output layer

Suitable activation function



## Justify:

Number of hidden neurons

Choice of activation function

Print  the total number of trainable parameters.


In [ ]:
import torch.nn as nn

# Get the number of input features from X_train_tensor
input_size = X_train_tensor.shape[1]

# Define the neural network model
class SimpleNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(SimpleNN, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()

        self.fc2 = nn.Linear(hidden_size, output_size)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        out = self.fc1(x)
        out = self.relu(out)
        out = self.fc2(out)
        out = self.sigmoid(out)
        return out

# Hyperparameters for the model
hidden_size = 64 # Number of neurons in the hidden layer
output_size = 1  # For binary classification (outputting a single probability)

# Instantiate the model
model = SimpleNN(input_size, hidden_size, output_size)

# Print the model architecture
print("Model Architecture:")
print(model)

# Print the total number of trainable parameters
total_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f"\nTotal number of trainable parameters: {total_params}")

Model Architecture:
SimpleNN(
  (fc1): Linear(in_features=102, out_features=64, bias=True)
  (relu): ReLU()
  (fc2): Linear(in_features=64, out_features=1, bias=True)
  (sigmoid): Sigmoid()
)

Total number of trainable parameters: 6657


## Write Answer 02:
- The neural network model consists of an input layer, one hidden layer with 64 neurons, and an output layer.

 - ReLU activation function is used in the hidden layer, and a Sigmoid activation is used in the output layer, which is suitable for binary classification tasks.
 - The total number of trainable parameters in this model is 6657.

# Question 03: [ Marks 10 ]

Choose an appropriate loss function.

Choose an optimizer.

<br>

Justify your choices based on:

Regression vs Classification

Nature of the dataset

In [ ]:
import torch.optim as optim
import torch.nn as nn

# Choose an appropriate loss function for binary classification
loss_function = nn.BCELoss()

# Choose an optimizer
optimizer = optim.Adam(model.parameters(), lr=0.001)

# print("Loss Function:",loss_function)
# print("Optimizer:", optimizer)

## Write Answer 03:

**Loss Function Choice: Binary Cross-Entropy Loss (nn.BCELoss)**

**Justification:**

1.  **Problem Type (Binary Classification):** The problem is a binary classification task, where the goal is to predict whether an individual has Class/ASD (1) or not (0). Binary Cross-Entropy Loss is specifically designed for such tasks.
2.  **Nature of the Dataset/Model Output:** Our neural network's output layer uses a Sigmoid activation function, which squashes the final output into a probability score between 0 and 1. Binary Cross-Entropy Loss measures the performance of a classification model whose output is a probability value between 0 and 1. It quantifies the difference between the predicted probabilities and the true binary labels. A lower BCE loss indicates a better-performing model.

**Optimizer Choice: Adam (Adaptive Moment Estimation)**

**Justification:**

1.  **General Performance:** Adam is a widely popular and robust optimization algorithm that combines the best aspects of two other popular optimizers: AdaGrad and RMSProp. It is known for its efficiency and good performance in a variety of deep learning tasks, including classification.
2.  **Adaptive Learning Rates:** Adam computes adaptive learning rates for each parameter, meaning it adjusts the step size for each weight and bias during training. This makes it less sensitive to the choice of the initial learning rate compared to traditional Stochastic Gradient Descent (SGD).
3.  **Computational Efficiency:** It is computationally efficient and requires little memory, making it a practical choice for training neural networks, especially on larger datasets or more complex models. Its ability to handle sparse gradients and non-stationary objectives makes it a strong general-purpose optimizer.

# Question 04: [ Marks 15 ]

## Implement a full training loop:

Forward pass

Loss computation

Backward pass

Parameter update

Gradient reset

### Requirements:

Train for at least 100 epochs.

Print loss every 10 epochs.

Store training loss history(You can pick your own Data Structure).

Explain clearly what happens in each step of the pipeline.

In [ ]:
epochs = 100
learning_rate = 0.001 # Defined in optimizer already
training_loss_history = []

for epoch in range(epochs):
    # 1. Forward pass: Compute predicted y by passing X to the model
    y_pred = model(X_train_tensor)

    # 2. Loss computation: Calculate the loss
    # Reshape y_train_tensor to match outputs shape if necessary
    loss = loss_function(y_pred,y_train_tensor.reshape(-1,1))

    # 3. Backward pass: Compute gradient of the loss with respect to model parameters
    optimizer.zero_grad()  # Gradient reset: Clear previous gradients
    loss.backward()        # Backpropagation: Compute gradients

    # 4. Parameter update: Adjust model parameters using the optimizer
    optimizer.step()

    # Store training loss history
    training_loss_history.append(loss.item())

    # Print loss every 10 epochs
    if (epoch + 1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}')

print("\nTraining complete.")
print(f"Final training loss: {training_loss_history[-1]:.4f}")

Epoch [10/100], Loss: 0.5696
Epoch [20/100], Loss: 0.4714
Epoch [30/100], Loss: 0.3840
Epoch [40/100], Loss: 0.3093
Epoch [50/100], Loss: 0.2484
Epoch [60/100], Loss: 0.1999
Epoch [70/100], Loss: 0.1620
Epoch [80/100], Loss: 0.1327
Epoch [90/100], Loss: 0.1101
Epoch [100/100], Loss: 0.0927

Training complete.
Final training loss: 0.0927


## Write Answer 04:

The training loop has successfully completed, and the loss has decreased over the 100 epochs, indicating that the model is learning. The final training loss achieved was **0.0907**.

Here's an explanation of what happens in each step of the training pipeline:

1.  **Forward Pass**: `outputs = model(X_train_tensor)`
    *   In this step, the input training data (`X_train_tensor`) is fed into the neural network model. The model performs a series of computations, including matrix multiplications and applying activation functions (ReLU in the hidden layer and Sigmoid in the output layer), to generate predictions (`outputs`). These `outputs` are the model's estimate of the probability that each sample belongs to the positive class.

2.  **Loss Computation**: `loss = loss_function(y_pred,y_train_tensor.reshape(-1,1))`
    *   After the forward pass, the calculated `outputs` (predictions) are compared against the true target values (`y_train_tensor`). The `loss_function` (Binary Cross-Entropy Loss) quantifies the difference between the predicted probabilities and the actual labels. A lower loss value indicates that the model's predictions are closer to the true labels.

3.  **Gradient Reset**: `optimizer.zero_grad()`
    *   Before performing backpropagation for the current iteration (epoch), it is essential to clear the gradients from the previous iteration. PyTorch, by default, accumulates gradients. If we don't reset them, the gradients would be the sum of gradients from all previous iterations, leading to incorrect and unstable parameter updates. This step ensures that we compute gradients solely based on the current batch's loss.

4.  **Backward Pass (Backpropagation)**: `loss.backward()`
    *   This is the core of training. Based on the computed loss, `loss.backward()` calculates the gradients of the loss with respect to every trainable parameter (weights and biases) in the model. This process uses the chain rule to determine how much each parameter contributed to the error, propagating the error signal backward through the network layers.

5.  **Parameter Update**: `optimizer.step()`
    *   Finally, the `optimizer` (Adam in this case) uses the gradients calculated in the backward pass to adjust the model's parameters. It moves the parameters in the direction that reduces the loss. The `learning_rate` (set to 0.001) determines the size of these steps. This iterative adjustment process is how the model learns to make more accurate predictions over time.

# Question 05: [ Marks 10 ]

## Evaluate the model on test data.

## For regression:

Report MSE and MAE


## For classification:

Report Accuracy

Compare training vs testing performance.

State whether the model is underfitting or overfitting.

In [ ]:
from sklearn.metrics import accuracy_score

# Set the model to evaluation mode
model.eval()

# Disable gradient calculations during inference
with torch.no_grad():
    # Make predictions on the test set
    y_pred_test = model(X_test_tensor)

    # Convert predictions (probabilities) to binary labels (0 or 1)
    # A threshold of 0.5 is commonly used for binary classification with sigmoid output
    y_pred_labels = (y_pred_test > 0.5).float()

    # Calculate accuracy on the test set
    # Ensure y_test is also squeezed to match dimensions if necessary
    accuracy = accuracy_score(y_test.cpu().numpy(), y_pred_labels.cpu().numpy())

print(f"Test Accuracy: {accuracy:.4f}")

Test Accuracy: 0.8936


## Write Answer 05:

**Test Accuracy: 0.8936 (approximately 89.36%)**

**Comparison of Training vs. Testing Performance:**

*   **Training Performance:** The final training loss was **0.0907**. A low Binary Cross-Entropy loss generally indicates that the model is performing very well on the training data, achieving a high training accuracy.
*   **Testing Performance:** The test accuracy is **0.8936**. This means the model correctly classified approximately 89.36% of the samples in the unseen test dataset.

**Assessment of Underfitting or Overfitting:**

Given the very low training loss, it is highly probable that the model achieved a very high accuracy on the training set. The test accuracy of 0.8936 is slightly lower than what the training performance suggests. This indicates a slight degree of **overfitting**.

**Explanation:**

*   **Overfitting** occurs when a model learns the training data too well, capturing noise and specific patterns that do not generalize to new, unseen data. In this case, the model has become highly optimized for the training examples, which might include some dataset-specific quirks. When presented with the test set, which contains different examples, its performance drops slightly because it hasn't learned the broader, generalizable patterns as effectively.
*   While there is an indication of overfitting, the difference between the presumed high training accuracy and the 89.36% test accuracy is not drastic. This suggests that the overfitting is not severe, and the model still performs reasonably well on new data.

# Question 06: [ Marks 20 ]

## Modify at least ONE of the following:

Learning rate

Number of hidden neurons

Number of epochs

### Train again and compare:

Convergence speed

Final performance

Explain how the change affected the model.

In [ ]:
# --- Modification for Question 06: Increase hidden neurons --- #

# 1. Define new hyperparameters
# Previous hidden_size = 64
new_hidden_size = 128 # Increased number of neurons in the hidden layer

# Re-instantiate the model with the new hidden_size
model_q6 = SimpleNN(input_size, new_hidden_size, output_size)

# Re-initialize the optimizer for the new model
optimizer_q6 = optim.Adam(model_q6.parameters(), lr=0.001)

# Store training loss history for the new run
training_loss_history_q6 = []

print(f"\nTraining model with {new_hidden_size} hidden neurons...")

# 2. Train the model with new hyperparameters (same training loop)
epochs = 100
for epoch in range(epochs):
    # Forward pass
    y_pred_q6 = model_q6(X_train_tensor)

    # Loss computation
    loss_q6 = loss_function(y_pred_q6, y_train_tensor.reshape(-1,1))

    # Backward pass and parameter update
    optimizer_q6.zero_grad()
    loss_q6.backward()
    optimizer_q6.step()

    training_loss_history_q6.append(loss_q6.item())

    if (epoch + 1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss_q6.item():.4f}')

print("\nTraining complete for Q6 modification.")
print(f"Final training loss (new hidden size): {training_loss_history_q6[-1]:.4f}")

# 3. Evaluate the new model on test data
model_q6.eval() # Set the new model to evaluation mode
with torch.no_grad():
    y_pred_test_q6 = model_q6(X_test_tensor)
    y_pred_labels_q6 = (y_pred_test_q6 > 0.5).float()
    accuracy_q6 = accuracy_score(y_test.cpu().numpy(), y_pred_labels_q6.cpu().numpy())

print(f"Test Accuracy (new hidden size): {accuracy_q6:.4f}")

# Store the results for comparison
previous_final_train_loss = training_loss_history[-1]
previous_test_accuracy = accuracy

current_final_train_loss = training_loss_history_q6[-1]
current_test_accuracy = accuracy_q6



Training model with 128 hidden neurons...
Epoch [10/100], Loss: 0.5142
Epoch [20/100], Loss: 0.3841
Epoch [30/100], Loss: 0.2815
Epoch [40/100], Loss: 0.2062
Epoch [50/100], Loss: 0.1534
Epoch [60/100], Loss: 0.1170
Epoch [70/100], Loss: 0.0918
Epoch [80/100], Loss: 0.0740
Epoch [90/100], Loss: 0.0611
Epoch [100/100], Loss: 0.0514

Training complete for Q6 modification.
Final training loss (new hidden size): 0.0514
Test Accuracy (new hidden size): 0.9007


## Write Answer 06:

**Modification:** The number of hidden neurons was increased from 64 to 128.

**Results Comparison:**

*   **Original Model (64 hidden neurons):**
    *   Final Training Loss: 0.0927
    *   Test Accuracy: 0.8936

*   **Modified Model (128 hidden neurons):**
    *   Final Training Loss: 0.0514
    *   Test Accuracy: 0.9007

**Explanation of how the change affected the model:**

1.  **Convergence Speed:** The model with 128 hidden neurons achieved a lower loss much faster and converged to a significantly lower final training loss (0.0514 vs. 0.0927) within the same 100 epochs. This indicates that the increased capacity allowed the model to learn the training data more efficiently.

2.  **Final Performance:** Both the training and testing performance improved with the increased number of hidden neurons. The final training loss decreased, and the test accuracy saw a slight but positive increase from 0.8936 to 0.9007.

**Impact of the Change:**

Increasing the number of hidden neurons from 64 to 128 enhanced the model's **capacity**. A model with more neurons can learn more complex patterns and represent more intricate relationships within the data. In this specific case:

*   **Improved Learning:** The lower final training loss suggests that the larger model was better able to fit the training data, capturing more nuances.
*   **Better Generalization:** Crucially, the slight improvement in test accuracy (from 0.8936 to 0.9007) indicates that this increased capacity led to better generalization to unseen data, rather than just overfitting. This implies that the original model (with 64 neurons) might have been slightly under-capacity for the given dataset, and providing more computational units allowed it to extract more relevant features without exacerbating overfitting.

In summary, the increase in hidden neurons allowed the model to learn a more sophisticated representation of the data, resulting in faster convergence and a modest improvement in overall predictive performance on both training and test sets.

# Question 07: [ Marks 20 ]


# Training Analysis

Answer the following:

Why must gradients be reset every epoch?

What happens if learning rate is too high?

What happens if learning rate is too small?

Why do we define layers inside the constructor (__init__) and not inside forward()?


## Write Answer 07:

### Why must gradients be reset every epoch?
In PyTorch (and other deep learning frameworks), gradients are accumulated by default. This means that if we don't explicitly reset them, the gradients from the current batch/epoch will be added to the gradients from previous batches/epochs. This accumulation would lead to incorrect gradient values, as the optimizer would be trying to update the model parameters based on an average gradient over multiple steps, rather than the gradient calculated specifically for the current batch's loss. `optimizer.zero_grad()` clears these accumulated gradients, ensuring that each parameter update is based solely on the gradients computed from the current forward and backward pass.

### What happens if the learning rate is too high?
If the learning rate is too high, the optimizer will take very large steps during parameter updates. This can cause several issues:
1.  **Overshooting the minimum**: The optimizer might jump over the optimal minimum of the loss function, leading to oscillations around the minimum or even diverging completely.
2.  **Instability**: The loss function can become unstable, showing erratic behavior, and may not converge at all.
3.  **Divergence**: In extreme cases, a very high learning rate can cause the model's weights to grow exponentially, leading to NaN (Not a Number) values in the loss or weights, and the training process will diverge.

### What happens if the learning rate is too small?
Conversely, if the learning rate is too small, the optimizer will take very tiny steps during parameter updates:
1.  **Slow convergence**: The training process will be very slow, taking a long time to reach the minimum of the loss function. This means the model might require many more epochs to train effectively.
2.  **Getting stuck in local minima**: The model might get trapped in a suboptimal local minimum or a saddle point, as the small steps might not be sufficient to escape flatter regions of the loss landscape.
3.  **Suboptimal performance**: Even if it converges, it might settle for a suboptimal solution, not achieving the best possible performance.

### Why do we define layers inside the constructor ( `__init__` ) and not inside `forward()`?
In PyTorch's `nn.Module`:
1.  **Initialization and State Management**: Defining layers (like `nn.Linear`, `nn.ReLU`, etc.) in the `__init__` method ensures that these layers are initialized once when the model object is created. They become part of the model's state, meaning their parameters (weights and biases) are registered as `nn.Parameter` objects. This allows PyTorch to automatically track these parameters for gradient computation during the backward pass and include them in the `model.parameters()` iterator for the optimizer.
2.  **Efficiency**: If layers were defined inside the `forward()` method, a new instance of each layer would be created every time the `forward()` method is called (i.e., for every single forward pass during training or inference). This would be extremely inefficient, as it would repeatedly allocate memory and re-initialize weights and biases. It would also prevent the model from learning, as the parameters would be reset in each step, making backpropagation and parameter updates meaningless.
3.  **Modularity and Readability**: It separates the definition of the model's architecture (what layers it has) from its computation logic (how data flows through those layers), improving code organization, readability, and maintainability. The `__init__` method sets up the building blocks, and `forward()` describes how these blocks are connected.